enterprise/server/sociartifactstore/sociartifactstore.go

```diff
@@ -285,112 +344,191 @@ func deserializeDigest(s string) (*repb.Digest, error) {
+	var annotations = map[string]string{
+		sociImageLayerDigestKey:    layerDigest,
+		sociImageLayerMediaTypeKey: layerMediaType,
+	}
+	return &ocispec.Descriptor{
+		MediaType:   octetStreamMediaType,
+		Digest:      godigest.NewDigestFromEncoded(godigest.SHA256, ztocDigest.Hash),
+		Size:        ztocDigest.SizeBytes,
+		Annotations: annotations,
+	}
 }
 
-func findSociIndex(ctx context.Context, imageRef string) (*repb.Digest, error) {
-	// TODO(iain): make the path a variable or something
-	db, err := soci.NewDB(sociDbPath)
+func imageDescriptor(image v1.Image) (*ocispec.Descriptor, error) {
+	mediaType, err := image.MediaType()
 	if err != nil {
 		return nil, err
 	}
-
-	var sociIndex *repb.Digest = nil
-	err = db.Walk(func(entry *soci.ArtifactEntry) error {
-		if entry.Type == soci.ArtifactEntryTypeIndex && entry.ImageDigest == imageRef {
-			sociIndex = &repb.Digest{
-				Hash:      strings.ReplaceAll(entry.Digest, "sha256:", ""),
-				SizeBytes: entry.Size,
-			}
-			return nil
-		}
-		return nil
-	})
+	digest, err := image.Digest()
 	if err != nil {
 		return nil, err
 	}
-	return sociIndex, nil
-}
-
-func toDigest(hash string) (*repb.Digest, error) {
-	stat, err := os.Stat(blobPath(hash))
+	size, err := image.Size()
 	if err != nil {
 		return nil, err
 	}
-	return &repb.Digest{
-		Hash:      hash,
-		SizeBytes: stat.Size(),
+	return &ocispec.Descriptor{
+		MediaType: string(mediaType),
+		Digest:    godigest.Digest(digest.String()),
+		Size:      size,
 	}, nil
 }
 
-func (s *SociArtifactStore) writeArtifactsToCache(ctx context.Context, imageId string, sociIndexDigest *repb.Digest, ztocDigests []*repb.Digest) error {
-	if _, err := s.blobstore.WriteBlob(ctx, blobKey(imageId), []byte(serializeDigest(sociIndexDigest))); err != nil {
-		return err
+func (s *SociArtifactStore) indexLayer(ctx context.Context, layer v1.Layer) (*repb.Digest, error) {
+	mediaType, err := layer.MediaType()
+	if err != nil {
+		return nil, err
 	}
-	if err := s.writeArtifactToCache(ctx, sociIndexDigest); err != nil {
-		return err
+	layerDigest, err := layer.Digest()
+	if err != nil {
+		return nil, err
 	}
-	for _, ztocDigest := range ztocDigests {
-		if err := s.writeArtifactToCache(ctx, ztocDigest); err != nil {
-			return err
-		}
+	layerSize, err := layer.Size()
+	if err != nil {
+		return nil, err
 	}
-	return nil
-}
 
-func (s *SociArtifactStore) writeArtifactToCache(ctx context.Context, digest *repb.Digest) error {
-	bytes, err := os.ReadFile(blobPath(digest.Hash))
+	compressionAlgo, err := images.DiffCompression(ctx, string(mediaType))
 	if err != nil {
-		return err
+		return nil, status.NotFoundErrorf("could not determine layer compression: %s", err)
+	}
+	if compressionAlgo != compression.Gzip {
+		return nil, status.UnimplementedErrorf("layer %s (%s) cannot be indexed because it is compressed with %s",
+			layerDigest.Hex, mediaType, compressionAlgo)
+	}
+
+	layerTmpFile, err := os.CreateTemp("", "layer.*")
```

Comment:
We should add a GKE volume (can use `emptyDir`) for this scratch space since layers can be large. The space on the host node is not unlimited and w/o setting up your own storage, GKE won't know how much you will need. 
You should change this code to take a flag to point to a scratch directory and then separately add a volume to the GKE deployment.

Reply:
Sounds good, done, with another PR coming for the new volume